libvmaf/src/framesync.c

```diff
@@ -0,0 +1,227 @@
+    void *frame_data;
+    int buf_status;
+    signed long index;
+    struct VmafFrameSyncBuf *next;
+} VmafFrameSyncBuf;
+
+typedef struct VmafFrameSyncContext {
+    VmafFrameSyncBuf *buf_que;
+    pthread_mutex_t acquire_lock;
+    pthread_mutex_t retrieve_lock;
+    pthread_cond_t retrieve;
+    unsigned buf_cnt;
+} VmafFrameSyncContext;
+
+int vmaf_framesync_init(VmafFrameSyncContext **fs_ctx)
+{
+    VmafFrameSyncContext *const ctx = *fs_ctx = malloc(sizeof(VmafFrameSyncContext));
+    if (!ctx) return -ENOMEM;
+    memset(ctx, 0, sizeof(VmafFrameSyncContext));
+    ctx->buf_cnt = 1;
+
+    pthread_mutex_init(&(ctx->acquire_lock), NULL);
+    pthread_mutex_init(&(ctx->retrieve_lock), NULL);
+    pthread_cond_init(&(ctx->retrieve), NULL);
+
+    VmafFrameSyncBuf *buf_que = ctx->buf_que = malloc(sizeof(VmafFrameSyncBuf));
+
+    buf_que->frame_data = NULL;
+    buf_que->buf_status = BUF_FREE;
+    buf_que->index = -1;
+    buf_que->next  = NULL;
+
+    return 0;
+}
+
+int vmaf_framesync_acquire_new_buf(VmafFrameSyncContext *fs_ctx, void **data,
+                                   unsigned data_sz, unsigned index)
+{
+    VmafFrameSyncBuf *buf_que = fs_ctx->buf_que;
+    *data = NULL;
+
+    pthread_mutex_lock(&(fs_ctx->acquire_lock));
+
+    // traverse until a free buffer is found
+    for (unsigned i = 0; i < fs_ctx->buf_cnt; i++) {
+        if (buf_que->buf_status == BUF_FREE) {
+            buf_que->frame_data = *data = malloc(data_sz);
+            if (!buf_que->frame_data)
+                return -ENOMEM;
+            buf_que->buf_status = BUF_ACQUIRED;
+            buf_que->index = index;
+            break;
+        }
+        // move to next node
+        if (buf_que->next != NULL)
+            buf_que = buf_que->next;
+    }
+
+    // create a new node if all nodes are occupied in the list and append to the tail
+    if (*data == NULL) {
+        VmafFrameSyncBuf *new_buf_node = malloc(sizeof(VmafFrameSyncBuf));
+        buf_que->next = new_buf_node;
+        new_buf_node->buf_status = BUF_FREE;
+        new_buf_node->index = -1;
+        new_buf_node->next = NULL;
+        fs_ctx->buf_cnt++;
+
+        new_buf_node->frame_data = *data = malloc(data_sz);
+        if (!new_buf_node->frame_data)
+            return -ENOMEM;
+        new_buf_node->buf_status = BUF_ACQUIRED;
+        new_buf_node->index = index;
+    }
+
+    pthread_mutex_unlock(&(fs_ctx->acquire_lock));
+
+    return 0;
+}
+
+int vmaf_framesync_submit_filled_data(VmafFrameSyncContext *fs_ctx, void *data,
+                                      unsigned index)
+{
+    VmafFrameSyncBuf *buf_que = fs_ctx->buf_que;
+
+    pthread_mutex_lock(&(fs_ctx->retrieve_lock));
+
+    // loop until a matchng buffer is found
+    for (unsigned i = 0; i < fs_ctx->buf_cnt; i++) {
+        if ((buf_que->index == index) && (buf_que->buf_status == BUF_ACQUIRED)) {
+            buf_que->buf_status = BUF_FILLED;
+            if (data != buf_que->frame_data)
```

Comment:
what is the purpose of this check? 

It seems that to succeed, the pair `(index, data)` needs to match the pair `(buf_que->index, buf_que->frame_data)`. Curious about:
1. Why that condition is needed
2. Why the treatment is different if the index doesn't match (do nothing) vs if the index matches but the data pointer doesn't (mark as filled and return -1)
3. What if the entire loop finishes without finding a match?

Reply:
Forwarded to Ittiam.

Reply:
Q1:
Index matching is done to find the node which holds the data buffer of picture "index"

Q2:
Index matching is the primary check to change the buffer status to FILLED, 
Data pointer checking is for error checking purpose, which can be cleaned up if not necessary. if the caller has not passed the matching data buffer, then there can be OOB or unallocated buffer access, this additional check is added to catch such cases.
Q3:
This condition is not applicable since the thread which has acquired new buffer for any particular picture index is responsible for submitting it after filling, so index should match with a node in the queue.

Reply:
Thanks @IttiamVijayakumarGR for clarifying, I was discussing this with @kylophone and it seemed like the extra check can be cleaned up, so I did that in the PR below. I will leave the last call to Kyle though.

Reply:
Discussed this with @nilfm99, we'll merge this one as-is for now to unblock the funque upstreaming work. However, I think it's likely that we'll have some small changes to the api in a later iteration. Since this is an internal api with just one use so far, we do have some flexibility when it comes to refactoring.